src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+            get_arrow_value::<arrow::datatypes::Float32Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_bits()))
+        }
+        CreateFunctionDataType::F64 | CreateFunctionDataType::DOUBLE => {
+            get_arrow_value::<arrow::datatypes::Float64Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_bits()))
+        }
+        CreateFunctionDataType::TEXT
+        | CreateFunctionDataType::CHAR
+        | CreateFunctionDataType::VARCHAR => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::StringArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to string array",
+                col_ix
+            ))),
+        },
+        CreateFunctionDataType::BOOLEAN => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::BooleanArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to boolean array",
+                col_ix
+            ))),
+        },
+        // serialize decimal as string
+        CreateFunctionDataType::DECIMAL {
+            precision: _,
+            scale: _,
+        } => get_arrow_value::<arrow::datatypes::Decimal128Type>(args, row_ix, col_ix)
+            .map(|val| Value::from(val.to_string())),
+        // dates are represented as i32 integer internally, serialize them as such.
+        CreateFunctionDataType::DATE => {
+            get_arrow_value::<arrow::datatypes::Date32Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_string()))
+        }
+        // timestamps are represented as i64 integers internally, serialize them as such.
+        CreateFunctionDataType::TIMESTAMP => get_arrow_value::<
+            arrow::datatypes::TimestampNanosecondType,
+        >(args, row_ix, col_ix)
+        .map(|val| Value::from(val.to_string())),
+    }
+}
+
+fn decode_udf_result_primitive_array<T>(
+    encoded_results: &Vec<Value>,
+    decoder: &dyn Fn(&Value) -> Result<T::Native>,
+) -> Result<PrimitiveArray<T>>
+where
+    T: ArrowPrimitiveType,
+{
+    let mut decoded_results = Vec::with_capacity(encoded_results.len());
+    for i in encoded_results {
+        decoded_results.push(Some(decoder(i)?));
+    }
+    Ok(decoded_results.iter().collect::<PrimitiveArray<T>>())
+}
+
+fn messagepack_decode_results(
+    return_type: &CreateFunctionDataType,
+    encoded_results: &Vec<Value>,
+) -> Result<ArrayRef> {
+    match return_type {
+        CreateFunctionDataType::SMALLINT => decode_udf_result_primitive_array::<
+            arrow::datatypes::Int16Type,
+        >(encoded_results, &|v| match v
+            .as_i64()
```

Comment:
Does messagepack not let you do `as_i16`/`as_i32` directly?

Reply:
No, not sure why. I think the reason might be that it has variable length integer encodings and the serialized size of the integer (which is input dependent) shouldn't leak into the types used in the code (which is checked compile-time). Instead, it's all hid behind `i64`.